### DISCUSSION TITLE
Optional bool in query string always True when parameter not present

### DISCUSSION THREAD

**Discussion by LKay** [fastapi/fastapi#8216] [Questions] [answered]
**Describe the bug**

When query parameter of type `Optional[bool]` is declared using `Query` class parameter value is not correctly mapped and instead of `None` it maps to instance of `Schema` which when type casted to `bool` returns True.

**To Reproduce**

Sample code:

```python
class QueryParams(BaseModel):
	foo: Optional[bool] = None

async def query_params(foo: Optional[bool] = Query(None, description="Some parameter")):
	return QueryParams(foo=foo)

@api.get("/foo")
async def get(data: QueryParams = Depends(query_params)):
	# When '/foo' is called (without any parameters) the 'foo' parameter 
  # inside 'data' has value of 'True' instead of 'None'.
	return None
```

It however works fine if declared as just:

```python
async def query_params(foo: Optional[bool] = None):
	return QueryParams(foo=foo)
```

**Expected behavior**
Value should be assigned only if parameter is present in query string, `None` otherwise.

**Environment:**
 - OS: macOS
 - FastAPI Version: 0.35
 - Python version: 3.7.3

**Comment by dmontagu** [comment]
@LKay 
I can't reproduce this; the following test passes for me on the current version:
```python
from typing import Optional

from pydantic import BaseModel
from starlette.testclient import TestClient

from fastapi import FastAPI, Depends, Query

app = FastAPI()


class QueryParams(BaseModel):
    foo: Optional[bool] = None


async def query_params(foo: Optional[bool] = Query(None, description="Some parameter")):
    return QueryParams(foo=foo)


@app.get("/foo")
async def get(data: QueryParams = Depends(query_params)):
    # When '/foo' is called (without any parameters) the 'foo' parameter
    # inside 'data' has value of 'True' instead of 'None'.
    return data.foo


def test_optional_bool():
    assert TestClient(app).get("/foo").content == b"null"
```
Did I miss something?

**Comment by LKay** [comment]
Oh, sorry I missed one more part. It happens when there is another wrapping function for parameters ie. for admin and non-admin parameters, such as:

```python

class QueryParams(BaseModel):
    foo: Optional[bool] = None
    bar: Optional[bool] = None # ie. only admin param

async def query_params_admin(foo: Optional[bool] = Query(None, description="Some parameter"), bar: Optional[bool] = Query(None, description="Some parameter")):
    return QueryParams(foo=foo, bar=bar)

async def query_params(foo: Optional[bool] = Query(None, description="Some parameter")):
    return await query_params_admin(foo)

```

**Comment by dmontagu** [comment, marked as answer]
The problem in your example is that you are calling `query_params_admin` directly -- FastAPI can only perform dependency injection if the function is called as a result of `Depends`. (Note: this is not a bug in fastapi.)

If you make a self-contained example like the one I posted above that isn't working the way you'd expect, I can show you options for how to refactor it into something that works.

**Comment by LKay** [comment]
I got it. I refactored my code so I now have two function, both returning instance of my parameters models class.

**Comment by tiangolo** [comment]
Thanks for the help @dmontagu ! :bowing_man: :rocket: 

Thanks for reporting back and closing the issue @LKay :tada:
